checks.d/consul.py

```diff
@@ -0,0 +1,215 @@
+
+    def consul_request(self, instance, endpoint):
+        url = "{0}{1}".format(instance.get('url'), endpoint)
+        try:
+            resp = requests.get(url)
+        except requests.exceptions.Timeout:
+            self.log.exception('Consul request to {0} timed out'.format(url))
+            raise
+
+        resp.raise_for_status()
+        return resp.json()
+
+    ### Consul Config Accessors
+    def _get_local_config(self, instance):
+        #TODO: Maybe store the config on `self` here so we don't have to keep retrieving it?
+        # But how do we invalidate?
+
+        return self.consul_request(instance, '/v1/agent/self')
+
+    def _get_cluster_leader(self, instance):
+        return self.consul_request(instance, '/v1/status/leader')
+
+    def _get_agent_url(self, instance):
+        self.log.debug("Starting _get_agent_url")
+        local_config = self._get_local_config(instance)
+        agent_addr = local_config.get('Config', {}).get('AdvertiseAddr')
+        agent_port = local_config.get('Config', {}).get('Ports', {}).get('Server')
+        agent_url = "{0}:{1}".format(agent_addr, agent_port)
+        self.log.debug("Agent url is %s" % agent_url)
+        return agent_url
+
+    def _get_agent_datacenter(self, instance):
+        local_config = self._get_local_config(instance)
+        agent_dc = local_config.get('Config', {}).get('Datacenter')
+        return agent_dc
+
+    ### Consul Leader Checks
+    def _is_instance_leader(self, instance):
+        try:
+            agent_url = self._get_agent_url(instance)
+            leader = self._last_known_leader or self._get_cluster_leader(instance)
+            self.log.debug("Consul agent lives at %s . Consul Leader lives at %s" % (agent_url,leader))
+            return agent_url == leader
+
+        except Exception as e:
+            return False
+
+    def _check_for_leader_change(self, instance):
+        agent_dc = self._get_agent_datacenter(instance)
+        leader = self._get_cluster_leader(instance)
+
+        if not leader:
+            # A few things could be happening here.
+            #   1. Consul Agent is Down
+            #   2. The cluster is in the midst of a leader election
+            #   3. The Datadog agent is not able to reach the Consul instance (network partition et al.)
+            self.log.warn('Consul Leader information is not available!')
+            return
+
+        if not self._last_known_leader:
+            # We have no state preserved, store some and return
+            self._last_known_leader = leader
+            return
+
+        if leader != self._last_known_leader:
+            self.log.info(('Leader change from {0} to {1}. Sending new leader event').format(
+                self._last_known_leader, leader))
+
+            self.event({
+                "timestamp": int(time.time()),
+                "event_type": "consul.new_leader",
+                "api_key": self.agentConfig.get("api_key", ''),
```

Comment:
that's not needed:
https://github.com/DataDog/dd-agent/blob/master/checks/__init__.py#L464-L465